src/wasm-type.h

```diff
@@ -164,12 +190,161 @@ struct Signature {
   }
   bool operator!=(const Signature& other) const { return !(*this == other); }
   bool operator<(const Signature& other) const;
+  std::string toString() const;
+};
+
+struct Field {
+  Type type;
```

Comment:
This might become a problem in that `StorageType` has a notion of `i8` and `i16`, differing from everything else, affecting both `Struct` and `Array`.

Interestingly, AssemblyScript has a notion of `i8`, `u8`, `i16` and `u16` and I once suggested to make these first-class types in Binaryen so we don't have to do the wrapping/sign-extension on our end. That'd be ideal from an AssemblyScript point of view while also solving this problem, but might be too much to ask for, in which case I'd expect that we'd hack around this only for `Field` types. Let me know what you think :)

Reply:
Interesting! I'll have to go read the proposal again to brush up on this. My feeling is that if field types are not the same as normal types, they should probably have a separate `FieldType` type, rather than trying to re-use normal `Type`s.

Reply:
The alternative here is to do the same as is done for `Load` and `Store`, with a `bytes` and `signed_` field. So the `Type` would be `i32`, with `bytes=1/2` and `signed=true/false` for the smaller types, otherwise these fields are irrelevant incl. for `i64` which is relevant for `Load` and `Store` but not for `Field` as it currently stands. (I still secretly hope to get first-class small types so asc doesn't have to go to lengths anymore to support them, but it's really a lot to ask for 😅)

Reply:
Hmm, how are field types used? Do the field types ever appear in the types of instructions? My guess is no, which is why I would suggest keeping them separate from the core `Type`s, even if there is some duplication there. (But I'll go brush up on the proposal now to make sure I know what I'm talking about).

Reply:
These are used by the [instructions](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#instructions-1)

Opcode | Type | Parameters
-|-|-
0xfb04 | `struct.get_s $t i` | `$t : typeidx`, `i : fieldidx`
0xfb05 | `struct.get_u $t i` | `$t : typeidx`, `i : fieldidx`

with `$t` being [defined](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#type-definitions) as `storagetype ::= <valtype> | <packedtype>` with `packedtype ::= i8 | i16`. Overall quite similar to `i32.load_8s` for example, where we are currently using `bytes` and `signed_` to indicate, implicitly emitting the right opcode. As such, implicitly emitting the right `storagetype` without an actual concept of it would be close to what other code already does.

Reply:
Oh, wait, I got this wrong. These are not part of the instruction, but as you suspected really just part of the `field` type definition's .

Reply:
Somewhat live in the same number space in between `v128` and `funcref`, but that's merely a detail of the binary format it seems and doesn't imply that these are used elsewhere.

Opcode | Type
-|-
-0x05 | v128
-0x06 | i8
-0x07 | i16
-0x10 | funcref

Reply:
Yeah I still think a separate `FieldType` makes sense, then. Arguably normal loads and stores would be cleaner if they had a similar `MemoryType`, but that's probably not worth working on.

Reply:
Ok, what about

```cpp
struct Field {
  enum FieldType {
    i32 = Type::i32,
    i64 = Type::i64,
    f32 = Type::f32,
    f64 = Type::f64,
    v128 = Type::v128,
    funcref = Type::funcref,
    externref = Type::externref,
    nullref = Type::nullref,
    exnref = Type::exnref,
    i8 = Type::_last_basic_id + 1,
    i16
  } type;
  ...
```

squeezing `i8` and `i16` in between `_last_basic_id` and the first constructed type?

Reply:
Ohh, I wasn't thinking about how `FieldType` could be any normal `Type` plus additionally `i8` or `i16`. I was incorrectly thinking about it as some different finite set of types. Sorry about that. You're totally right that we should be using `Type` in this case.

What if we did

```C++
struct FieldType {
    enum FieldTypeKind { I8, I16, Other } kind;
    Type type; // Type::none if kind == I8 or kind == I16
};
```

This is so gross in C++ and would be so simple in Rust :'(

Reply:
Heh, building upon your suggestion, what about

```cpp
struct Field {
  Type type;
  enum Packed { not_packed, i8, i16 } packed; // applicable iff type=i32
  bool mutable_;
```

Reply:
Looks good to me!

Reply:
Here's [the diff in isolation](https://github.com/WebAssembly/binaryen/pull/3012/commits/12183ecd3c5441675ea190e14ce47669e9240d91#diff-0c49fb11df8bd2e482f4e71293448677R207-R208), with a comment :)